verifier/providerverifier.go

```diff
@@ -0,0 +1,277 @@
+	return v.issuer
+}
+
+func (v *DefaultProviderVerifier) VerifyProvider(ctx context.Context, pkt *pktoken.PKToken) error {
+	// Check whether Audience claim matches provided Client ID
+	// No error is thrown if option is set to skip client ID check
+	if err := verifyAudience(pkt, v.options.ClientID); err != nil && !v.options.SkipClientIDCheck {
+		return err
+	}
+
+	alg, ok := pkt.ProviderAlgorithm()
+	if !ok {
+		return fmt.Errorf("provider algorithm type missing")
+	}
+
+	if alg != gq.GQ256 && v.options.GQOnly {
+		return ErrNonGQUnsupported
+	}
+
+	switch alg {
+	case gq.GQ256:
+		if err := VerifyGQSig(ctx, pkt); err != nil {
+			return fmt.Errorf("error verifying OP GQ signature on PK Token: %w", err)
+		}
+	case jwa.RS256:
+		opToken, err := pkt.Compact(pkt.Op)
+		if err != nil {
+			return err
+		}
+
+		pubKey, err := v.ProviderPublicKey(ctx, opToken)
+		if err != nil {
+			return fmt.Errorf("failed to get OP public key: %w", err)
+		}
+
+		token, err := pkt.Compact(pkt.Op)
+		if err != nil {
+			return err
+		}
+
+		if _, err := jws.Verify(token, jws.WithKey(alg, pubKey)); err != nil {
+			return err
+		}
+	}
+
+	if err := v.verifyCommitment(pkt); err != nil {
+		return err
+	}
+
+	return nil
+}
+
+// This function takes in an OIDC Provider created ID token or GQ-signed modification of one and returns
+// the associated public key
+func (v *DefaultProviderVerifier) ProviderPublicKey(ctx context.Context, token []byte) (crypto.PublicKey, error) {
+	message, err := jws.Parse(token)
+	if err != nil {
+		return nil, err
+	}
+
+	// a JWT is guaranteed to have exactly one signature
+	headers := message.Signatures()[0].ProtectedHeaders()
+
+	alg, ok := headers.Get(jws.AlgorithmKey)
+	if !ok {
+		return nil, fmt.Errorf("missing algorithm header")
+	}
+
+	var providerHeaders jws.Headers
+	if alg == gq.GQ256 {
+		origHeaders, err := originalTokenHeaders(token)
+		if err != nil {
+			return nil, fmt.Errorf("malformatted PK token headers: %w", err)
+		}
+
+		if origHeaders.Algorithm() != jwa.RS256 {
+			return nil, fmt.Errorf("expected original headers to contain RS256 alg, got %s", headers.Algorithm())
+		}
+
+		providerHeaders = origHeaders
+	} else {
+		providerHeaders = headers
+	}
+
+	// Extract our issuer from the payload claims
+	var claims struct {
+		Issuer string `json:"iss"`
+	}
+	if err := json.Unmarshal(message.Payload(), &claims); err != nil {
+		return nil, err
+	}
+
+	return v.options.DiscoverPublicKey(ctx, providerHeaders, claims.Issuer)
+}
+
+func (v *DefaultProviderVerifier) verifyCommitment(pkt *pktoken.PKToken) error {
+	var claims map[string]any
+	if err := json.Unmarshal(pkt.Payload, &claims); err != nil {
+		return err
+	}
+
+	cic, err := pkt.GetCicValues()
+	if err != nil {
+		return err
+	}
+	expectedCommitment, err := cic.Hash()
+	if err != nil {
+		return err
+	}
+	commitment, ok := claims[v.commitmentClaim]
+	if !ok {
+		return fmt.Errorf("missing commitment claim %s", v.commitmentClaim)
+	}
+
```

Comment:
Added a clarifying comment here.

```suggestion

// The OIDC Spec allows the aud (audience) claim to be list or a string. 
// In the case of an aud commitment claim, we assume it is a string and 
// we do not handle the case where it is a list.
```

We might want to add a check if we get a list for the aud commitment and output an error message explaining we encountered an aud commitment claim that is a list. Not a blocker, can be done in a future PR.